fhir/client.go

```diff
@@ -413,3 +449,176 @@ func DiscardAndClose(r io.ReadCloser) error {
+// The error is formatted using the given format with all potential substitutions.
+func DownloadBundleError(format string, a ...interface{}) DownloadBundle {
+	return DownloadBundle{
+		Err: fmt.Errorf(format, a...),
+	}
+}
+
+func (c *Client) ExpandPages(initialRequest *http.Request, resChannel chan<- DownloadBundle) {
+	var requestStart time.Time
+	var processingStart time.Time
+	var request = initialRequest
+	var nextPageURL *url.URL
+	var err error
+
+	for ok := true; ok; ok = nextPageURL != nil {
+		var stats networkStats
+
+		if nextPageURL != nil {
+			request, err = c.NewPaginatedRequest(nextPageURL)
+		}
+		if err != nil {
+			resChannel <- DownloadBundleError("could not create FHIR server request: %v\n", err)
+			return
+		}
+
+		trace := &httptrace.ClientTrace{
+			GotConn: func(_ httptrace.GotConnInfo) {
+				requestStart = time.Now()
+			},
+			WroteRequest: func(_ httptrace.WroteRequestInfo) {
+				processingStart = time.Now()
+			},
+			GotFirstResponseByte: func() {
+				stats.ProcessingDuration = time.Since(processingStart).Seconds()
+			},
+		}
+		request = request.WithContext(httptrace.WithClientTrace(request.Context(), trace))
+
+		response, err := c.Do(request)
+		if err != nil {
+			resChannel <- DownloadBundleError("could not request the FHIR server with URL %s: %v\n", request.URL, err)
+			return
+		}
+
+		responseBody, err := io.ReadAll(response.Body)
+		if err != nil {
+			resChannel <- DownloadBundleError("could not read FHIR server response after request to URL %s: %v\n", request.URL, err)
+			return
+		}
+		if err := response.Body.Close(); err != nil {
+			resChannel <- DownloadBundleError("could not close the response body: %v\n", err)
+			return
+		}
+		stats.RequestDuration = time.Since(requestStart).Seconds()
+		stats.TotalBytesIn += int64(len(responseBody))
+
+		if response.StatusCode != http.StatusOK {
+			outcome, err := fm.UnmarshalOperationOutcome(responseBody)
+			if err != nil {
+				bundle := DownloadBundleError("request to FHIR server with URL %s had a non-ok response status (%d) but the expected operation outcome could not be parsed: %v", request.URL, response.StatusCode, err)
+				bundle.Stats = &stats
+				resChannel <- bundle
+				return
+			}
+
+			bundle := DownloadBundleError("request to FHIR server with URL %s had a non-ok response status (%d)", request.URL, response.StatusCode)
+			bundle.ErrResponse = &util.ErrorResponse{
+				StatusCode:       response.StatusCode,
+				OperationOutcome: &outcome,
+			}
+			bundle.Stats = &stats
+			resChannel <- bundle
+			return
+		}
+
+		nextPageURL, err = extractNextLink(response, resChannel, request, responseBody)
+		if err != nil {
+			return
+		}
+
+		resChannel <- DownloadBundle{
+			AssociatedRequestURL: *request.URL,
+			ResponseBody:         responseBody,
+			Stats:                &stats,
+		}
+	}
+}
+
+func extractNextLink(response *http.Response, resChannel chan<- DownloadBundle, request *http.Request, responseBody []byte) (*url.URL, error) {
```

Comment:
I would not extract this method because it needs too many arguments, especially the channel. I would like to keep the channel in one function